python/cudf/cudf/core/index.py

```diff
@@ -2991,89 +2990,56 @@ def interval_range(
-
-    if periods and not freq:
-        # if statement for mypy to pass
-        if end is not None and start is not None:
-            # divmod only supported on host side scalars
-            quotient, remainder = divmod((end - start).value, periods.value)
-            if remainder:
-                freq_step = cudf.Scalar((end - start) / periods)
-            else:
-                freq_step = cudf.Scalar(quotient)
-            if start.dtype != freq_step.dtype:
-                start = start.astype(freq_step.dtype)
-            bin_edges = sequence(
-                size=periods + 1,
-                init=start.device_value,
-                step=freq_step.device_value,
-            )
-            left_col = bin_edges.slice(0, len(bin_edges) - 1)
-            right_col = bin_edges.slice(1, len(bin_edges))
-    elif freq and periods:
-        if end:
-            start = end - (freq * periods)
-        if start:
-            end = freq * periods + start
-        if end is not None and start is not None:
-            left_col = arange(
-                start.value, end.value, freq.value, dtype=common_dtype
-            )
-            end = end + 1
-            start = start + freq
-            right_col = arange(
-                start.value, end.value, freq.value, dtype=common_dtype
-            )
-    elif freq and not periods:
-        if end is not None and start is not None:
-            end = end - freq + 1
-            left_col = arange(
-                start.value, end.value, freq.value, dtype=common_dtype
-            )
-            end = end + freq + 1
-            start = start + freq
-            right_col = arange(
-                start.value, end.value, freq.value, dtype=common_dtype
-            )
-    elif start is not None and end is not None:
-        # if statements for mypy to pass
-        if freq:
-            left_col = arange(
-                start.value, end.value, freq.value, dtype=common_dtype
-            )
-        else:
-            left_col = arange(start.value, end.value, dtype=common_dtype)
-        start = start + 1
-        end = end + 1
-        if freq:
-            right_col = arange(
-                start.value, end.value, freq.value, dtype=common_dtype
-            )
-        else:
-            right_col = arange(start.value, end.value, dtype=common_dtype)
-    else:
-        raise ValueError(
-            "Of the four parameters: start, end, periods, and "
-            "freq, at least two must be specified"
-        )
+
+    periods = periods.astype("int64")
+    common_dtype = find_common_type((start.dtype, freq.dtype, end.dtype))
+    start = start.astype(common_dtype)
+    freq = freq.astype(common_dtype)
+
+    bin_edges = sequence(
+        size=periods + 1,
+        init=start.device_value,
+        step=freq.device_value,
+    )
+    left_col = bin_edges.slice(0, len(bin_edges) - 1)
+    right_col = bin_edges.slice(1, len(bin_edges))
```

Comment:
You're doing `linspace` here with `arange`-like calls, are we sure the edge cases are handled correctly (specifically if `periods` has not been provided).

Reply:
Ah ok, they are also wrong then:

```
In [10]: pd.interval_range(start=1, end=3.0, freq=0.5)
Out[10]: IntervalIndex([(1.0, 1.5], (1.5, 2.0], (2.0, 2.5], (2.5, 3.0]], dtype='interval[float64, right]')

In [11]: pd.interval_range(start=1, end=3.0, freq=0.1)
Out[11]: IntervalIndex([(1.0, 1.1055555555555556], (1.1055555555555556, 1.211111111111111], (1.211111111111111, 1.3166666666666667], (1.3166666666666667, 1.4222222222222223], (1.4222222222222223, 1.5277777777777777] ... (2.3722222222222222, 2.477777777777778], (2.477777777777778, 2.583333333333333], (2.583333333333333, 2.688888888888889], (2.688888888888889, 2.7944444444444443], (2.7944444444444443, 2.9]], dtype='interval[float64, right]')
```

Reply:
Note how I asked for a frequency of 0.1 (so the intervals should be `(1, 1.1]` etc...)

Reply:
Ah, because the implementation does loads of stuff with floats that suffers from rounding errors

Reply:
Yup should hopefully be fixed in 2.1 in pandas to just do an arange in these cases https://github.com/pandas-dev/pandas/pull/54477